### PR TITLE
musl libunwind test

### PR DESCRIPTION
**Pull request by satmandu**
- Attempt to work around musl i686 issues by adding libunwind to linking.
- Only enabled for `i686`

Works properly:
- [x] x86_64
- [x] i686 ???

### Run the following to test this pull request locally on i686.
```
CREW_TESTING_REPO=https://github.com/satmandu/chromebrew.git CREW_TESTING_BRANCH=musl_lzip CREW_TESTING=1 crew update && \
crew install -s musl_libunwind ; \
crew upgrade lzip ; \
lzip --version || (crew remove lzip && crew install -s lzip)
```

### CODE REVIEW

File: packages/lzip.rb [hunk 1]
```diff
@@ -3,32 +3,73 @@
...
   source_sha256 'c3342d42e67139c165b8b128d033b5c96893a13ac5f25933190315214e87a948'
 
   binary_url ({
-    aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22_armv7l/lzip-1.22-chromeos-armv7l.tar.xz',
-     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22_armv7l/lzip-1.22-chromeos-armv7l.tar.xz',
-       i686: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22_i686/lzip-1.22-chromeos-i686.tar.xz',
-     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22_x86_64/lzip-1.22-chromeos-x86_64.tar.xz',
+    aarch64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22-1_armv7l/lzip-1.22-1-chromeos-armv7l.tpxz',
+     armv7l: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22-1_armv7l/lzip-1.22-1-chromeos-armv7l.tpxz',
+     x86_64: 'https://gitlab.com/api/v4/projects/26210301/packages/generic/lzip/1.22-1_x86_64/lzip-1.22-1-chromeos-x86_64.tpxz',
```
Comment (uberhacker) [inline packages/lzip.rb:15]: Binaries still on 1.22-1?

File: packages/musl_libbacktrace.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+require 'package'
+
+class Musl_libbacktrace < Package
+  description 'Library to produce symbolic backtraces'
+  homepage 'https://github.com/ianlancetaylor/libbacktrace'
+  version 'd0f5e95a87a4d3e0a1ed6c069b5dae7cbab3ed2a'
+  compatibility 'all'
+  source_url 'https://github.com/ianlancetaylor/libbacktrace.git'
+  git_hashtag version
+
```
Comment (uberhacker) [inline packages/musl_libbacktrace.rb:13]: No binaries?

File: packages/lzip.rb (hunk 1 above)
Comment (satmandu) [reply packages/lzip.rb:15]: This is for testing... for you on i686... no binaries yet for that reason.

File: packages/musl_libbacktrace.rb (hunk 2 above)
Comment (satmandu) [reply packages/musl_libbacktrace.rb:13]: this is entirely for testing on i686, and I'm not going to generate binaries for it...

**Comment by uberhacker** [comment]
Unable to build on i686.  See below:
```
chronos@localhost ~/Downloads $ crew build lzip
lzip: Lzip is a lossless data compressor with a user interface similar to the one of gzip or bzip2.
The following packages also need to be installed: 
musl_libunwind musl_libbacktrace musl_libucontext 
Do you agree? [Y/n] 
Proceeding...
musl_libunwind: libunwind is a portable and efficient C programming interface (API) to determine the call-chain of a program.
No precompiled binary available for your platform, downloading source...
##################################################################################################################################### 100.0%
Musl_libunwind archive downloaded.
Unpacking archive using 'tar', this may take a while...
Building from source, this may take a while...
Rename all *.la files to *.la_tmp
checking build system type... i686-pc-linux-gnu
checking host system type... i686-pc-linux-gnu
checking target system type... i686-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for gcc... /usr/local/musl/bin/i686-linux-musl-gcc
checking whether the C compiler works... no
configure: error: in `/usr/local/tmp/crew/musl_libunwind.20211121183219.dir/libunwind-1.5.0':
configure: error: C compiler cannot create executables
See `config.log' for more details
musl_libunwind failed to build: `./configure --prefix=/usr/local/musl       PATH=/usr/local/musl/bin:/home/chronos/user/.config/composer/vendor/bin:/usr/local/opam/default/bin:/usr/local/opam/default/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/home/chronos/user/go/bin       CC='/usr/local/musl/bin/i686-linux-musl-gcc'       CXX='/usr/local/musl/bin/i686-linux-musl-g++'       LD=/usr/local/musl/bin/i686-linux-musl-ld.gold       LIBDIR=/usr/local/musl/lib       INCLUDEDIR=/usr/local/musl/include       PKG_CONFIG_LIBDIR=/usr/local/musl/lib/pkgconfig       PKGCONFIGDIR=/usr/local/musl/lib/pkgconfig       CFLAGS='-B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic-interposition -fdevirtualize-at-ltrans  -fcommon'       CXXFLAGS='-B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic-interposition -fdevirtualize-at-ltrans  -fcommon -static'       CPPFLAGS='-I/usr/local/musl/include -fcommon'       LDFLAGS='-L/usr/local/musl/lib -flto -static'       --enable-shared       --enable-static       --enable-ptrace       --disable-tests` exited with 77
```

**Comment by uberhacker** [comment]
Relevant config.log details below:
```
configure:3928: checking whether the C compiler works
configure:3950: /usr/local/musl/bin/i686-linux-musl-gcc -B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic$
cc1: fatal error: conftest.c: No such file or directory
compilation terminated.
```

**Comment by satmandu** [comment]
> Relevant config.log details below:
> ```
> configure:3928: checking whether the C compiler works
> configure:3950: /usr/local/musl/bin/i686-linux-musl-gcc -B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic$
> cc1: fatal error: conftest.c: No such file or directory
> compilation terminated.
> ```

I feel like you have seen this before.... Configure fails to write a `conftest.c` to disk...  Can you mount /use/local/tmp/ from an external disk (maybe via USB?) and see if the results are the same?

It still feels like the sort of disk issue one would get from not being able to write to a disk.

**Comment by uberhacker** [comment]
> > Relevant config.log details below:
> > ```
> > configure:3928: checking whether the C compiler works
> > configure:3950: /usr/local/musl/bin/i686-linux-musl-gcc -B/usr/local/musl/include -flto -pipe -O3 -ffat-lto-objects -fipa-pta -fno-semantic$
> > cc1: fatal error: conftest.c: No such file or directory
> > compilation terminated.
> > ```
> 
> I feel like you have seen this before.... Configure fails to write a `conftest.c` to disk... Can you mount /use/local/tmp/ from an external disk (maybe via USB?) and see if the results are the same?
> 
> It still feels like the sort of disk issue one would get from not being able to write to a disk.

It's more sinister than that.  I simply get a segfault running `/usr/local/musl/bin/i686-linux-musl-gcc`.

**Comment by satmandu** [comment]
Just in case the chromebrew package is corrupted, can you tell me if this works:
```
mkdir /tmp/musl ; curl -Lsk https://musl.cc/i686-linux-musl-native.tgz | tar -C /tmp/musl -zxvf -
/tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc --version
```

That gives me this:
```
/tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc --version
i686-linux-musl-gcc (GCC) 10.2.1 20210227
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Also FYI:
```
sha256sum /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
ceaaee4e7d28755adff30eb90ea65f8110bf7937ccb7facdf33ba745d87df812  /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
```

I'd be curious to know what this gives:
```
sha256sum /usr/local/musl/bin/i686-linux-musl-gcc
```

**Comment by satmandu** [comment]
Also in my container:
```
sha256sum /usr/local/musl/bin/i686-linux-musl-gcc
ceaaee4e7d28755adff30eb90ea65f8110bf7937ccb7facdf33ba745d87df812  /usr/local/musl/bin/i686-linux-musl-gcc
```

**Comment by uberhacker** [comment]
There is a difference.  See below:
```
$ sha256sum /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
908344fb0689f4b20a9c6aef0ff7ab22962d8cae32dc599b6206fb2f795b0996  /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
$ sha256sum /usr/local/musl/bin/i686-linux-musl-gcc
ceaaee4e7d28755adff30eb90ea65f8110bf7937ccb7facdf33ba745d87df812  /usr/local/musl/bin/i686-linux-musl-gcc
```
Must be the stripping.

**Comment by satmandu** [comment]
> There is a difference.  See below:
> ```
> $ sha256sum /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
> 908344fb0689f4b20a9c6aef0ff7ab22962d8cae32dc599b6206fb2f795b0996  /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
> $ sha256sum /usr/local/musl/bin/i686-linux-musl-gcc
> ceaaee4e7d28755adff30eb90ea65f8110bf7937ccb7facdf33ba745d87df812  /usr/local/musl/bin/i686-linux-musl-gcc
> ```
> Must be the stripping.

The package version has the correct hash from the output you shared...

Let me extract that source tgz outside the container and check the hash.  So weird.

**Comment by satmandu** [comment]
From ubuntu running `( mkdir /tmp/musl ; curl -Lsk https://musl.cc/i686-linux-musl-native.tgz | tar -C /tmp/musl -zxvf - ; /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc --version ; sha256sum /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc )` I get this:

```
i686-linux-musl-gcc (GCC) 11.2.1 20211120
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

908344fb0689f4b20a9c6aef0ff7ab22962d8cae32dc599b6206fb2f795b0996  /tmp/musl/i686-linux-musl-native/bin/i686-linux-musl-gcc
```

**Comment by uberhacker** [comment]
Yeah, they both work for me too so it's mystery what is causing the segfault.

**Comment by satmandu** [comment]
I just realized that the musl toolchains got updated! So let's see if the new version doesn't segfault?

Can you see if the extracted into /tmp musl-gcc segfaults? i.e.
```
( rm -rf /tmp/musl ; mkdir /tmp/musl ; curl -Lsk https://musl.cc/i686-linux-musl-native.tgz | tar --strip-components=1  -C /tmp/musl -zxvf - ; /tmp/musl/bin/i686-linux-musl-gcc --version )
```

If that gives this:
```
i686-linux-musl-gcc (GCC) 11.2.1 20211120
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Then you could try overwriting the packaged install of `musl_native_toolchain` thus:
```
( curl -Lsk https://musl.cc/i686-linux-musl-native.tgz | tar --strip-components=1  -C /usr/local/musl -zxvf - ; /usr/local/musl/bin/i686-linux-musl-gcc --version )
```

Followed by an attempt at `crew build lzip`

File: packages/lzip.rb (hunk 1 above)
Comment (satmandu) [reply packages/lzip.rb]: added binaries after building using newer musl...

**Comment by satmandu** [comment]
ok I updated the toolchain.

Let's see if any of this works?

If this doesn't work either... Ugh... I have other ideas.

**Comment by uberhacker** [comment]
Works!
```
$ lzip --version
lzip 1.22
Copyright (C) 2021 Antonio Diaz Diaz.
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by uberhacker** [comment]
Building lzip from source works on i686 too.  Nice work!

**Pull request merged**